sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/streaming/StreamingReadSupport.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.reader.streaming;
+
+import org.apache.spark.sql.sources.v2.reader.ReadSupport;
+
+/**
+ * A base interface for streaming read support. This is package private and is invisible to data
+ * sources. Data sources should implement concrete streaming read support interfaces:
+ * {@link MicroBatchReadSupport} or {@link ContinuousReadSupport}.
+ */
+interface StreamingReadSupport extends ReadSupport {
+
+  /**
+   * Returns the initial offset for a streaming query to start reading from. Note that the
+   * streaming data source should not assume that it will start reading from its
+   * {@link #initialOffset()} value: if Spark is restarting an existing query, it will restart from
+   * the check-pointed offset rather than the initial one.
+   */
+  Offset initialOffset();
+
+  /**
+   * Deserialize a JSON string into an Offset of the implementation-defined offset type.
+   *
+   * @throws IllegalArgumentException if the JSON does not encode a valid offset for this reader
+   */
+  Offset deserializeOffset(String json);
+
+  /**
+   * Informs the source that Spark has completed processing all data for offsets less than or
+   * equal to `end` and will only request offsets greater than `end` in the future.
+   */
+  void commit(Offset end);
```

Comment:
I think this should accept a `ScanConfig`. The read support is general and can create multiple scans. It should not keep state about any one scan. That's something the `ScanConfig` should do.

Reply:
I'd +1 passing a ScanConfig. I agree that all the existing sources are just going to pull out the offset, but "Spark is finished with this scan" is a cleaner semantic than "Spark is finished with the scan such that it goes up to this offset".

Reply:
In Kafka, isn't the ScanConfig what identifies the consumer group? @jose-torres

Reply:
Not in the current implementation.

Reply:
Why not? A `ScanConfig` represents the scan. The thing that is scanned shouldn't store the scan state.

Reply:
I'm not sure what you mean by "scan state" here. The thing that is scanned needs to know what offsets are available for scanning, which requires holding a Kafka consumer to read that information.

Reply:
I tried it but this can't work for continuous streaming.

The `ScanConfig` in continuous streaming represents the scan from a certain offset, but it doesn't know when it ends. It's the streaming engine that tracks the offset we have committed and call this method.

Reply:
I don't understand what you mean. A scan should be associated with a consumer or consumer group. The offset that is being updated is for that consumer, not the topic. And because the ReadSupport object corresponds to the topic in this abstraction (the thing that can be scanned), it shouldn't be what tracks the consumer or consumer group.

All we need to do to fix the API is to pass in ScanConfig here.

@cloud-fan, I'm not sure what you mean that this doesn't work. Why would tracking the consumer in the ScanConfig for a continuous streaming job not work?

Reply:
There are two consumer groups in streaming:

1. The one at the driver, which determines what offsets are available to scan.
2. The one distributed across the executors which actually performs the scan.

This method is used to commit certain offsets in group 1, based on the offsets which have been logged as processed by group 2. In microbatch mode, this happens to work with ScanConfig, because there is one ScanConfig for each offset log entry. In continuous mode there is one ScanConfig corresponding to an indefinite number of offset log entries, so ScanConfig does not provide the information required to commit any particular entry.

Reply:
> Why would tracking the consumer in the ScanConfig for a continuous streaming job not work?

It's spark (the streaming engine) that tracks the consumers, not streaming source. If you insist to pass a `ScanConfig`, the only way I can think of is to make `ScanConfig` mutable, then Spark sets the end offset to the `ScanConfig` and calls this `commit`. This is a really bad API design and we should not do it. I'd like to keep this `commit` method as it is.

Reply:
Okay, I think I see where the misunderstanding is coming from: the current Spark implementation keeps track of offsets itself and doesn't commit those offsets in an external system. That makes sense and I would not want Spark to make the ScanConfig mutable to update it with the scan's current offset.

However, as a streaming API I think that it is entirely possible for an implementation to commit offsets to an external system based on a scan, so it makes sense to me to pass ScanConfig in for that purpose, when a ScanConfig does represent a single consumer of a scan.